### DISCUSSION TITLE
defaultValue propagation in Controller

### DISCUSSION THREAD

**Discussion by vertic4l** [react-hook-form/react-hook-form#7273] [Q&A]
Given the following Controller setup:

```tsx
<Controller
            control={props.control}
            name={props.name}
            render={({ field: { onChange, value } }) => {
                return (
                    <CheckboxGroup
                        {...props}
                        onChange={(val) => {
                            onChange(val);
                            if (props.onChange) {
                                props.onChange(val);
                            }
                        }}
                        value={value}
                    />
                );
            }}
        />
```

We set defaultValues here:

```tsx
const {
        handleSubmit,
        register,
        control,
        formState: { errors },
    } = useForm<FormData>({
        mode: "onSubmit",
        reValidateMode: "onBlur",
        resolver: yupResolver(schema),
        defaultValues: (formData
            ? toJS(formData.data)
            : {}) as UseFormProps<FormData>["defaultValues"],
    });
```

value of the Controller's render function doesn't get the initial defaultValue, but after triggering onChange it always gives the correct value as it should be.

Is this intended ?

non controlled inputs for example are getting the default value immediately.

I would consider that the value comes from the defaultValues of useForm and delivers the value on first render.

Or is it a bug ?

**Comment by vertic4l** [comment]
<img width="913" alt="Bildschirmfoto 2021-12-13 um 12 28 09" src="https://user-images.githubusercontent.com/1771763/145804555-850ab0c6-740c-4e81-a072-dbf8a7d4c5f8.png">


"The defaultValues for inputs are used as the initial value when a component is first rendered"

"...the value from defaultValues will be used."


I really believe that this is a bug.

@bluebill1049

**Comment by bluebill1049** [comment]
> defaultValues are cached on the first render within the custom hook. If you want to reset the defaultValues, you should use the reset API.

https://react-hook-form.com/api/useform

**Comment by vertic4l** [comment]
@bluebill1049 

```tsx
useEffect(() => {
        reset();
        // eslint-disable-next-line react-hooks/exhaustive-deps
    }, []);
```

still empty value. Wrong timing ? And even if its cached, if it gets initialized with defaultValues, why are these empty in controller ?

**Comment by Southclaws** [comment]
I think I'm running into the same issue, if I log out the result of `watch()` on first render, the values are not there but my validation somehow passes (which I'm using the zod resolver for and all fields are required).

If I remove the defaultValues from useForm, the validation starts to work again.

There seems to be some disconnect between the default values being set, and the validation/rendering. My form is empty on fresh render, but the defaults have been set properly. If I submit the form, the values aren't there but the validation is not doing anything.

My setup is <Controller> too for some custom NativeBase form elements:

```tsx
      <Controller
        control={control}
        name={name}
        render={({ field }) => (
          <Input
            {...props}
            value={field.value}
            defaultValue={ /* there's nothing from the render props that fits here */ }
            onChangeText={(v) => field.onChange(v)}
            autoCompleteType={autoCompleteType ?? "off"}
          />
        )}
      />
```

I think the issue is there's no field in the render props to pass to the `defaultValue` prop of the enclosed input component.

My workaround for now is simply to use `setValue` on the first render:

```
  useEffect(() => {
    // check if form untouched
    // setValue("fieldname", defaultvalue) for all fields
   }
```

**Reply by Moshyfawn** [reply]
Can you reproduce in [this CSB](https://codesandbox.io/s/magical-snowflake-ix1j3?file=/src/components/Form.tsx)?
